tensorflow/lite/kernels/activations.cc

```diff
@@ -870,14 +1006,12 @@ TfLiteStatus SigmoidEval(TfLiteContext* context, TfLiteNode* node) {
     }
     case kTfLiteInt16: {
       LogisticParams params;
-      if (kernel_type == kReference) {
-        reference_ops::Logistic(
-            params, GetTensorShape(input), GetTensorData<int16_t>(input),
-            GetTensorShape(output), GetTensorData<int16_t>(output));
-      } else {
+      if (kernel_type == kFixedPointOptimized) {
         optimized_ops::Logistic(
             params, GetTensorShape(input), GetTensorData<int16_t>(input),
             GetTensorShape(output), GetTensorData<int16_t>(output));
+      } else {
+        EvalUsingLookupTableSigmoid16Bit(data, input, output);
```

Comment:
Just to make sure I'm understanding this correctly, this changes TFLite's builtin sigmoid ("kGenericOptimized") from optimized_ops::Logistic() to "EvalUsingLookupTableSigmoid16Bit", which is more accurate. And "EvalUsingLookupTableSigmoid16Bit" is not vectorized yet (like in "EvalUsingLookupTable").

Will "EvalUsingLookupTableSigmoid16Bit" be optimized in the future?

Reply:
As with other operators with 16-bit activations, our initial plans are to introduce the reference implementations. We have tentative plans for optimized implementations. Our other relevant efforts are currently focused on implementing corresponding kernels in TensorFlow Lite Micro, but that work depends on the progress with introducing the reference code to TFLite first.

Reply:
Thanks, I was not in some of the discussions so I might miss context here. Isn't it safer to change just the reference code and leave the two optimized unchanged?

Reply:
The issue with keeping the old kGenericOptimized implementation is the fact that it is the one currently used as the default BuiltinOp (see [Register_TANH()](https://github.com/tensorflow/tensorflow/blob/38eeb4f5d18c6772886b1f41093f4681bb522108/tensorflow/lite/kernels/activations.cc#L1342)). One of the incentives for this PR is to be able to run models quantized with higher accuracy provided by 16-bit activations using the default/standard TFLite build.

I would suggest that we enable the proposed code by default, and address any performance issues in due course. The alternative approach could be using kReference implementation by default, but, since it would affect execution paths for other, potentially more widely used, data types, I think it may not be the best option.

Reply:
Thanks Anton. I see your point and that makes a lot of sense. The only concern I have is that there are already model that runs with 16bit builtin Tanh/Sigmoid Op. This change will make it more accurate but in the same time slower, which can be an issue.

May I suggest the following? We can submit the code but only enable it in the reference case. Once the optimization is finished, we can flip the switch so existing users can get better accuracy without performance impact.

Does that sound good? Thanks!

Reply:
Thanks Jian Li. I appreciate your concerns, and generally agree with the proposed course of action. One thing that I would like to do before making the mods you are asking for is to move the new reference implementation to the [reference_integer_ops](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/kernels/internal/reference/integer_ops/tanh.h) namespace. It would make it possible to re-use the reference implementation in the code external to the TFLite run-time.

What do you think?

Reply:
Thanks for understanding the restrictions.

We all appreciate the improved accuracy this PR brings. What you mentioned works and please go ahead with the plan (either here or in another PR, whichever you feel comfortable with).

Thanks!

Reply:
Thanks for understanding the restrictions.

We all appreciate the improved accuracy this PR brings. What you mentioned works and please go ahead with the plan (either here or in another PR, whichever you feel comfortable with).

Thanks!